### PR TITLE
[SPARK-31935][SQL] Hadoop file system config should be effective in data source options

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Mkae Hadoop file system config effective in data source options.

From `org.apache.hadoop.fs.FileSystem.java`:
```
  public static FileSystem get(URI uri, Configuration conf) throws IOException {
    String scheme = uri.getScheme();
    String authority = uri.getAuthority();

    if (scheme == null && authority == null) {     // use default FS
      return get(conf);
    }

    if (scheme != null && authority == null) {     // no authority
      URI defaultUri = getDefaultUri(conf);
      if (scheme.equals(defaultUri.getScheme())    // if scheme matches default
          && defaultUri.getAuthority() != null) {  // & default has authority
        return get(defaultUri, conf);              // return default
      }
    }
    
    String disableCacheName = String.format("fs.%s.impl.disable.cache", scheme);
    if (conf.getBoolean(disableCacheName, false)) {
      return createFileSystem(uri, conf);
    }

    return CACHE.get(uri, conf);
  }
```
Before changes, the file system configurations in data source options are not propagated in `DataSource.scala`.
After changes, we can specify authority and URI schema related configurations for scanning file systems.

This problem only exists in data source V1. In V2, we already use `sparkSession.sessionState.newHadoopConfWithOptions(options)` in `FileTable`.
### Why are the changes needed?

Allow users to specify authority and URI schema related Hadoop configurations for file source reading.

### Does this PR introduce _any_ user-facing change?

Yes, the file system related Hadoop configuration in data source option will be effective on reading.

### How was this patch tested?

Unit test

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @liancheng @gatorsmile

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 1]
```diff
@@ -570,7 +570,7 @@ case class DataSource(
       checkEmptyGlobPath: Boolean,
       checkFilesExist: Boolean): Seq[Path] = {
     val allPaths = caseInsensitiveOptions.get("path") ++ paths
-    val hadoopConf = sparkSession.sessionState.newHadoopConf()
+    val hadoopConf = sparkSession.sessionState.newHadoopConfWithOptions(options)
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:573]: There are multiple `newHadoopConf`s in this file. Should we also fix them?

Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:576]: Yes, we should. I am adding more test cases.

**Comment by liancheng** [comment]
LGTM, thanks!

**Review by HyukjinKwon — APPROVED** [review]
Main code changes look good. LGTM if tests are fixed and pass.

**Review by cloud-fan — APPROVED** [review]
LGTM. How far shall we backport it?

**Comment by gengliangwang** [comment]
@cloud-fan I think we can backport it to branch-2.4. Once branch-3.0 rc3 is cut, we can backport it to branch-3.0 as well.
WDYT?

**Comment by gengliangwang** [comment]
retest this please

**Comment by gengliangwang** [comment]
Merging to master.
I will backport to branch-2.4 and branch-3.0 later

**Comment by dongjoon-hyun** [comment]
Hi, All.
The new test case seems flaky. This fails consistently in `Hadoop 3.2 / Hive 2.3`.
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-sbt-hadoop-3.2-hive-2.3/789/testReport/org.apache.spark.sql/FileBasedDataSourceSuite/SPARK_31935__Hadoop_file_system_config_should_be_effective_in_data_source_options/history/

Screen Shot 2020-06-10 at 4 22 12 PM

**Comment by dongjoon-hyun** [comment]
I'll make a follow-up PR.

**Pull request closed without merging**